Allow custom 'msgtype's for MessageEventContent

oppiliappan opened the pull request:
- [x] write custom `Deserialize` implementation
 - [x] work without `serde(tag)`
 - [x] include `_Custom(...)` variant in `MessageEventContent`
 ~Public API / Accessor methods~
~Tests~

Implements most of #406.

oppiliappan committed:
write custom `Deserialize` implementation for `MessageEventContent`

 - move serde(rename) attribute from enum variants to individual types
 - introduce MessageDeHelper to help with deserialization of
   MessageEventContent

jplatte commented on ruma-events/src/lib.rs:
```diff
@@ -520,6 +521,18 @@ pub struct EventDeHelper {
+
+    /// Everything else in the json object
+    #[serde(flatten)]
+    pub remaining: HashMap<String, JsonValue>,
```
```suggestion
    pub remaining: BTreeMap<String, JsonValue>,
```

We never use `HashMap` because iteration / serialization order is not deterministic there (plus perf. is worse for small amounts of entries).

oppiliappan replied:
Ah yeah, I forgot about this. What about `serde_json::Map`?

jplatte replied:
`serde_json::Map` only works if the key type is exactly `String`, for other types even if they are strings under the hood (e.g. `RoomId`) it doesn't implement any traits (or methods). So to be consistent we use `BTreeMap` everywhere (except in a single place). Also `serde_json::Map` is just a wrapper around `BTreeMap` (unless a certain feature is activated on `serde_json`).

jplatte replied:
I guess since we're wrapping `JsonValue` it might be slightly different here. Let me think and look at the existing code again...

jplatte commented on ruma-events/src/room/message.rs:
```diff
@@ -21,57 +21,48 @@ pub use super::relationships::InReplyTo;
 
 /// The payload for `MessageEvent`.
-#[derive(Clone, Debug, Deserialize, Serialize, MessageEventContent)]
+#[derive(Clone, Debug, MessageEventContent, Serialize)]
```
Please don't reorder these. They are ordered the same way as our imports in most places. (std ones first, then external, then internal)

jplatte commented on ruma-events/src/room/message.rs:
```diff
@@ -157,8 +148,34 @@ impl MessageEventContent {
+        D: de::Deserializer<'de>,
+    {
+        use MessageEventContent::*;
+        let json = Box::<RawJsonValue>::deserialize(deserializer)?;
```
Now that I see it here, I wonder whether we this would work instead:

```suggestion
        let json = <&RawJsonValue>::deserialize(deserializer)?;
```

oppiliappan replied:
`cargo test` failed with:

```
---- edit_deserialization_061 stdout ----
thread 'edit_deserialization_061' panicked at 'called `Result::unwrap()` on an
`Err` value: Error("invalid type: string
\"{\\\"body\\\":\\\"s/foo/bar\\\",\\\"m.new_content\\\":{\\\"body\\\":\\\"bar\\\"},\\\"m.relates_to\\\":{\\\"event_id\\\":\\\"$1598361704261elfgc:localhost\\\",\\\"rel_type\\\":\\\"m.replace\\\"},\\\"msgtype\\\":\\\"m.text\\\"}\",
expected raw value", line: 0, column: 0)',
ruma-events/tests/room_message.rs:148:59
```

jplatte replied:
What a useless error message. I remember some issues with `serde_json::Value` deserialization, maybe this is that. That resolves this concern in any case.

jplatte replied:
Yeah we have `BTreeMap<String, JsonValue>` in a few places. Let's go with that for now, if you want you can open a new issue about switching all these over at once.

oppiliappan committed:
make changes as requested

 - use BTreeMap over HashMap
 - rearrange derives

oppiliappan committed:
add _Custom variant to MessageEventContent

jplatte commented on ruma-events/src/room/message.rs:
```diff
@@ -157,8 +152,34 @@ impl MessageEventContent {
     }
 }
 
+impl<'de> de::Deserialize<'de> for MessageEventContent {
```
Any reason you didn't go with my suggestion from #406 to put this into a sub-module?

oppiliappan replied:
No reason really, I intended to factor it out into a new module once I got everything else done, in any case, I made a new module with the latest commit.

jplatte commented on ruma-events/src/lib.rs:
```diff
@@ -520,6 +520,18 @@ pub struct EventDeHelper {
+/// Helper struct to determine the msgtype from a `serde_json::value::RawValue`
+#[doc(hidden)]
+#[derive(Debug, Deserialize)]
+pub struct MessageDeHelper {
```
This type should be in `room::message` (or `room::message::content_serde`) and not be public. The fields also don't need `pub` then.

jplatte commented on ruma-events/src/room/message.rs:
```diff
@@ -585,3 +615,15 @@ pub struct KeyVerificationRequestEventContent {
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct CustomEventContent {
+    /// A custom msgtype
+    pub msgtype: String,
```
No need for the fields in this type to be `pub`lic.

oppiliappan replied:
These are public for the time being (for tests), Ill make these private as soon as the public API is ready.

jplatte commented on ruma-events/tests/room_message.rs:
```diff
@@ -73,6 +74,52 @@ fn content_serialization() {
+    let json_data = vec![("custom_field", "baba"), ("another_one", "abab")]
+        .into_iter()
+        .map(|(k, v)| (k.to_string(), json!(v)))
+        .collect();
```
You can use `maplit::btreemap` for this (maplit is a dev-dependency already).

oppiliappan committed:
refactor Deserialize implementation into content_serde module

 - use `btreemap!` in tests
 - move helper struct MessageDeHelper into content_serde

jplatte commented on ruma-events/src/room/message.rs:
```diff
@@ -18,10 +18,10 @@ use super::{relationships::RelatesToJsonRepr, EncryptedFile, ImageInfo, Thumbnai
 // FIXME: Do we want to keep re-exporting this?
 pub use super::relationships::InReplyTo;
 
+pub mod content_serde;
```
```suggestion
#[doc(hidden)]
pub mod content_serde;
```

jplatte commented on ruma-events/src/room/message.rs:
```diff
@@ -585,3 +590,15 @@ pub struct KeyVerificationRequestEventContent {
+
+    /// Remaining event content
+    #[serde(flatten)]
+    pub data: serde_json::Map<String, JsonValue>,
```
I guess I might have suggested this but after reviewing our other uses I think we can just use `BTreeMap` here too.

jplatte commented on ruma-events/src/room/message/content_serde.rs:
```diff
@@ -0,0 +1,50 @@
+            #[cfg(feature = "unstable-pre-spec")]
+            "m.key.verification.request" => VerificationRequest(from_raw_json_value(&json)?),
+            s => {
+                let remaining = remaining.into_iter().collect::<serde_json::Map<_, _>>();
```
... which would make this unnecessary.

jplatte commented on ruma-events/src/room/message/content_serde.rs:
```diff
@@ -0,0 +1,50 @@
+/// Helper struct to determine the msgtype from a `serde_json::value::RawValue`
+#[doc(hidden)]
+#[derive(Debug, Deserialize)]
+pub struct MessageDeHelper {
```
```suggestion
struct MessageDeHelper {
```

jplatte reviewed (commented):
I'd like to merge this as soon as possible so #420 can be rebased on top. Can you review the suggestions below? Public API and additional tests can be part of a follow-up PR then.

oppiliappan committed:
use BTreeMap over serde_json::Map in CustomEventContent

 - make changes as requested

jplatte commented:
This needs a rebase to fix the Nightly CI, also git history could be cleaned up. I'll do it for you in this case, for the future I highly recommend looking up interactive rebasing and fixup commits 🙂

oppiliappan commented:
I have rebased against upstream and cleaned up some of the git history:

```
342344df (HEAD -> custom-msgtypes) use BTreeMap over serde_json::Map in CustomEventContent
3c91a1ac refactor Deserialize implementation into content_serde module
092592ee add _Custom variant to MessageEventContent
8b7d33a6 write custom `Deserialize` implementation for `MessageEventContent`
317b2055 (upstream/main, main) events: Fix tests for unstable-pre-spec
```
If it is alright with you, I can force push this, but feel free to do as you please.

jplatte commented:
I'm already almost done. I also made some small adjustments that I'll note on the PR I'll post in a few minutes.

jplatte commented:
Continued in #423.
